### PR TITLE
add support for symlinked directories in private devel space

### PR DESCRIPTION
**Pull request by dominiquehunziker**
### Build / Run Issue

Symlinks in the package devel space linking to directories are not properly forwarded in the shared linked devel space layout, i.e., an empty directory is created instead of a symlink.

_Use case:_ I symlink a data directory to the devel space [like this](https://github.com/dominiquehunziker/catkin_tools_test_ws/blob/master/src/py_dummy_pkg/CMakeLists.txt#L9-L15) to make use of package data for Python packages in the devel space [here](https://github.com/dominiquehunziker/catkin_tools_test_ws/blob/master/src/py_dummy_pkg/src/py_dummy_pkg/loader.py#L4).

Sample workspace: https://github.com/dominiquehunziker/catkin_tools_test_ws

```
. /opt/ros/indigo/setup.sh
catkin clean -y
catkin config --link-devel
catkin build
. devel/setup.bash
rosrun py_dummy_pkg runner.py
```
### Suggested fix

Basically, the code/logic for files is reused (maybe this should be refactored?). However, I'm uncertain if the check for collisions is sufficient and if they should be handled differently from the files.

### CODE REVIEW

**Comment by remod** [comment]
I have the same issue. An easy fix which works for me is to replace

`os.walk(source_devel_path)`

with

`os.walk(source_devel_path, followlinks=True)`

Instead of cloning the symlink this will symlink its content.

**Comment by dominiquehunziker** [comment]
Yes, that works also.
However, the behavior is not exactly the same. In your case you would have to rerun catkin if you add a file to the symlinked directory. In my case this is transparent and, therefore, reflects the Python behavior in the devel space better.

**Comment by NikolausDemmel** [comment]
This change looks good to me. Haven't tested though.

**Comment by wjwwood** [comment]
I just tested this and I was able to reproduce the issue and see that this pr fixes the issue. The code changes are pretty involved, but they make sense to me. @jbohren if you ever have time, it would be good for you to review this change too. However, I'm going to merge it for now as it seems to work and not break anything I can see.

Thanks @dominiquehunziker and everyone else who chipped in on the discussion. Sorry it took so long to merge.
